src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectSnapshotManager.cs

```diff
@@ -63,6 +61,28 @@ internal partial class ProjectSnapshotManager(
     // we want to make sure the "add" finishes running first before "open" is notified.
     private readonly Queue<ProjectChangeEventArgs> _notificationQueue = new();
 
+    /// <summary>
+    /// Constructs an instance of <see cref="ProjectSnapshotManager"/>.
+    /// </summary>
+    /// <param name="projectEngineFactoryProvider">The <see cref="IProjectEngineFactoryProvider"/> to
+    /// use when creating <see cref="ProjectState"/>.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use.</param>
+    /// <param name="initializer">An optional callback to set up the initial set of projects and documents.
+    /// Note that this is called during construction, so it does not run on the dispatcher and notifications
+    /// will not be sent.</param>
+    public ProjectSnapshotManager(
+        IProjectEngineFactoryProvider projectEngineFactoryProvider,
+        ILoggerFactory loggerFactory,
+        Action<Updater>? initializer = null)
```

Comment:
This seems a little overcomplicated. Should this just take in a misc project instance and add if it's not null? 

That would prevent having the `_initialized` check as well

Reply:
Personally I'd be wary of avoiding using the `Updater`, and I think that is the thing that is pushing the need for `_initialized`.